core/common/src/main/java/alluxio/collections/NonUniqueFieldIndex.java

```diff
@@ -98,4 +111,80 @@ public T getFirst(Object value) {
     Set<T> all = mIndexMap.get(value);
     return all == null ? null : Iterables.getFirst(all, null);
   }
+
+  @Override
+  public Iterator<T> iterator() {
+    return new NonUniqueFieldIndexIterator();
+  }
+
+  @Override
+  public int size() {
+    int totalSize = 0;
+    for (ConcurrentHashSet<T> innerSet : mIndexMap.values()) {
+      totalSize += innerSet.size();
+    }
+    return totalSize;
+  }
+
+  /**
+   * Specialized iterator for {@link NonUniqueFieldIndex}.
+   *
+   * This is needed to support consistent removal from the set and the indices.
+   */
+  private class NonUniqueFieldIndexIterator implements Iterator<T> {
+    /**
+     * Iterator of {@link NonUniqueFieldIndex#mIndexMap}. This Iterator keeps track of the
+     * inner set which is under iteration.
+     */
+    private final Iterator<ConcurrentHashSet<T>> mIndexMapIterator;
+    /**
+     * Iterator inside each inner set. This iterator keeps track of the objects.
+     */
+    private Iterator<T> mIndexIterator;
+    /**
+     * Keeps track of current object. It is used to do the remove.
+     */
+    private T mObject;
+
+    public NonUniqueFieldIndexIterator() {
+      mIndexMapIterator = mIndexMap.values().iterator();
+      mIndexIterator = null;
+      mObject = null;
+    }
+
+    @Override
+    public boolean hasNext() {
```

Comment:
Yes, the `hasNext` simply checks the state. It will only update the outer iterator without changing the inner one. Thus it can be called multiple times and returns same value.

For the `next`, it can update the outer iterator by itself, so its okay to call `next` before the `hasNext`.